File: apps/kmp/composeApp/src/commonMain/kotlin/com/ticketinglabs/client/ui/screens/Screens.kt
package com.ticketinglabs.client.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import com.ticketinglabs.client.core.UiState
import com.ticketinglabs.client.domain.model.Event
import com.ticketinglabs.client.domain.model.EventDetail
import com.ticketinglabs.client.domain.model.Order
import com.ticketinglabs.client.domain.model.QueueToken
import com.ticketinglabs.client.domain.model.Reservation
import com.ticketinglabs.client.domain.model.Sector
import com.ticketinglabs.client.ui.atoms.CountdownTimer
import com.ticketinglabs.client.ui.atoms.MutedText
import com.ticketinglabs.client.ui.atoms.PrimaryButton
import com.ticketinglabs.client.ui.atoms.ScreenTitle
import com.ticketinglabs.client.ui.atoms.SecondaryButton
import com.ticketinglabs.client.ui.components.EventCard
import com.ticketinglabs.client.ui.components.OrderStatusPanel
import com.ticketinglabs.client.ui.components.PriceTag
import com.ticketinglabs.client.ui.components.QueuePositionCard
import com.ticketinglabs.client.ui.components.SectorRow
import com.ticketinglabs.client.ui.components.UiStateContent
import com.ticketinglabs.client.ui.theme.Tokens

/** A screen scaffold: title header over a scrollable, padded content column. */
@Composable
fun ScreenScaffold(title: String, content: @Composable () -> Unit) {
    Column(
        Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(Tokens.spaceXl),
        verticalArrangement = Arrangement.spacedBy(Tokens.spaceLg),
    ) {
        ScreenTitle(title)
        content()
    }
}

/** 1. Events list. */
@Composable
fun EventsScreen(state: UiState<List<Event>>, onOpen: (Event) -> Unit, onRetry: () -> Unit) {
    ScreenScaffold("Events") {
        UiStateContent(state, onRetry = onRetry, emptyText = "No events on sale.") { events ->
            Column(verticalArrangement = Arrangement.spacedBy(Tokens.spaceMd)) {
                events.forEach { EventCard(it, onOpen) }
            }
        }
    }
}

/** 2. Event detail. */
@Composable
fun EventDetailScreen(state: UiState<EventDetail>, onJoinQueue: () -> Unit, onRetry: () -> Unit) {
    ScreenScaffold("Event") {
        UiStateContent(state, onRetry = onRetry) { detail ->
            Column(verticalArrangement = Arrangement.spacedBy(Tokens.spaceMd)) {
                Text(detail.event.name, fontWeight = FontWeight.Bold)
                MutedText(detail.event.venue)
                detail.sectors.forEach { s ->
                    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                        Text(s.name)
                        PriceTag(s.price.amountCents, s.price.currency)
                    }
                }
                PrimaryButton("Join the queue", onJoinQueue)
            }
        }
    }
}

/** 3. Waiting room. */
@Composable
fun WaitingRoomScreen(state: UiState<QueueToken>, onContinue: () -> Unit, onRetry: () -> Unit) {
    ScreenScaffold("Waiting room") {
        UiStateContent(state, onRetry = onRetry) { token ->
            Column(verticalArrangement = Arrangement.spacedBy(Tokens.spaceLg)) {
                QueuePositionCard(token)
                if (token.isAdmitted) PrimaryButton("Choose seats", onContinue)
            }
        }
    }
}

/** 4. Sector selection with a quantity stepper; the CTA disables when sold out. */
@Composable
fun SectorSelectionScreen(detail: EventDetail, onReserve: (Sector, Int) -> Unit) {
    ScreenScaffold("Choose a sector") {
        Column(verticalArrangement = Arrangement.spacedBy(Tokens.spaceMd)) {
            detail.sectors.forEach { sector ->
                var quantity by remember(sector.id.value) { mutableStateOf(1) }
                Column(verticalArrangement = Arrangement.spacedBy(Tokens.spaceSm)) {
                    SectorRow(sector) { onReserve(sector, quantity) }
                    if (!sector.isSoldOut) {
                        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Tokens.spaceMd)) {
                            SecondaryButton("-", { if (quantity > 1) quantity-- })
                            Text("$quantity")
                            SecondaryButton("+", { if (quantity < 8) quantity++ })
                        }
                    }
                }
            }
        }
    }
}

/** 5. Reservation with a live expiry countdown; the CTA is disabled while not held. */
@Composable
fun ReservationScreen(
    state: UiState<Reservation>,
    remainingMs: Long,
    onCheckout: () -> Unit,
    onRetry: () -> Unit,
) {
    ScreenScaffold("Your hold") {
        UiStateContent(state, onRetry = onRetry) { reservation ->
            Column(verticalArrangement = Arrangement.spacedBy(Tokens.spaceLg)) {
                Text("${reservation.quantity} seat(s) held")
                MutedText("Complete checkout before the hold expires:")
                CountdownTimer(remainingMs)
                PrimaryButton("Checkout", onCheckout, enabled = reservation.isHeld && remainingMs > 0)
            }
        }
    }
}

/** 6 + 7. Checkout kicks off the order; this screen shows its live, polled status. */
@Composable
fun OrderStatusScreen(state: UiState<Order>, onDone: () -> Unit, onRetry: () -> Unit) {
    ScreenScaffold("Order") {
        UiStateContent(state, onRetry = onRetry) { order ->
            OrderStatusPanel(order, onDone)
        }
    }
}

File: apps/flutter/lib/ui/screens.dart
/// The seven flow screens. Stateless where possible: they take a [UiState] and callbacks
/// and render via [UiStateView].
library;

import 'package:flutter/material.dart';

import '../core/core.dart';
import '../domain/models.dart';
import 'theme.dart';
import 'widgets.dart';

Widget _scaffold(String title, Widget body) => Scaffold(
      appBar: AppBar(title: Text(title)),
      body: SingleChildScrollView(padding: const EdgeInsets.all(Tokens.spaceXl), child: body),
    );

/// 1. Events list.
class EventsScreen extends StatelessWidget {
  final UiState<List<Event>> state;
  final void Function(Event) onOpen;
  final VoidCallback onRetry;
  const EventsScreen({required this.state, required this.onOpen, required this.onRetry, super.key});
  @override
  Widget build(BuildContext context) => _scaffold(
        'Events',
        UiStateView<List<Event>>(
          state,
          onRetry: onRetry,
          emptyText: 'No events on sale.',
          onSuccess: (events) => Column(
            children: [
              for (final e in events) ...[EventCard(e, onOpen: () => onOpen(e)), const SizedBox(height: Tokens.spaceMd)],
            ],
          ),
        ),
      );
}

/// 2. Event detail.
class EventDetailScreen extends StatelessWidget {
  final UiState<EventDetail> state;
  final VoidCallback onJoinQueue;
  final VoidCallback onRetry;
  const EventDetailScreen({required this.state, required this.onJoinQueue, required this.onRetry, super.key});
  @override
  Widget build(BuildContext context) => _scaffold(
        'Event',
        UiStateView<EventDetail>(
          state,
          onRetry: onRetry,
          onSuccess: (detail) => Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(detail.event.name, style: const TextStyle(fontWeight: FontWeight.bold, fontSize: 18)),
              MutedText(detail.event.venue),
              const SizedBox(height: Tokens.spaceMd),
              for (final s in detail.sectors)
                Padding(
                  padding: const EdgeInsets.symmetric(vertical: Tokens.spaceXs),
                  child: Row(mainAxisAlignment: MainAxisAlignment.spaceBetween, children: [
                    Text(s.name),
                    PriceTag(s.price.amountCents, s.price.currency),
                  ]),
                ),
              const SizedBox(height: Tokens.spaceLg),
              PrimaryButton('Join the queue', onPressed: onJoinQueue),
            ],
          ),
        ),
      );
}

/// 3. Waiting room.
class WaitingRoomScreen extends StatelessWidget {
  final UiState<QueueToken> state;
  final VoidCallback onContinue;
  final VoidCallback onRetry;
  const WaitingRoomScreen({required this.state, required this.onContinue, required this.onRetry, super.key});
  @override
  Widget build(BuildContext context) => _scaffold(
        'Waiting room',
        UiStateView<QueueToken>(
          state,
          onRetry: onRetry,
          onSuccess: (token) => Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              QueuePositionCard(token),
              if (token.isAdmitted) ...[
                const SizedBox(height: Tokens.spaceLg),
                PrimaryButton('Choose seats', onPressed: onContinue),
              ],
            ],
          ),
        ),
      );
}

/// 4. Sector selection with a per-sector quantity stepper; the CTA disables when sold out.
class SectorSelectionScreen extends StatelessWidget {
  final EventDetail detail;
  final void Function(Sector sector, int quantity) onReserve;
  const SectorSelectionScreen({required this.detail, required this.onReserve, super.key});
  @override
  Widget build(BuildContext context) => _scaffold(
        'Choose a sector',
        Column(children: [for (final s in detail.sectors) _SectorPicker(s, onReserve)]),
      );
}

class _SectorPicker extends StatefulWidget {
  final Sector sector;
  final void Function(Sector, int) onReserve;
  const _SectorPicker(this.sector, this.onReserve);
  @override
  State<_SectorPicker> createState() => _SectorPickerState();
}

class _SectorPickerState extends State<_SectorPicker> {
  int quantity = 1;
  @override
  Widget build(BuildContext context) => Padding(
        padding: const EdgeInsets.only(bottom: Tokens.spaceMd),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            SectorRow(widget.sector, onSelect: () => widget.onReserve(widget.sector, quantity)),
            if (!widget.sector.isSoldOut)
              Row(children: [
                SecondaryButton('-', onPressed: () => setState(() => quantity = quantity > 1 ? quantity - 1 : 1)),
                Padding(padding: const EdgeInsets.symmetric(horizontal: Tokens.spaceMd), child: Text('$quantity')),
                SecondaryButton('+', onPressed: () => setState(() => quantity = quantity < 8 ? quantity + 1 : 8)),
              ]),
          ],
        ),
      );
}

/// 5. Reservation with a live expiry countdown.
class ReservationScreen extends StatelessWidget {
  final UiState<Reservation> state;
  final int remainingMs;
  final VoidCallback onCheckout;
  final VoidCallback onRetry;
  const ReservationScreen(
      {required this.state, required this.remainingMs, required this.onCheckout, required this.onRetry, super.key});
  @override
  Widget build(BuildContext context) => _scaffold(
        'Your hold',
        UiStateView<Reservation>(
          state,
          onRetry: onRetry,
          onSuccess: (reservation) => Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text('${reservation.quantity} seat(s) held'),
              const SizedBox(height: Tokens.spaceSm),
              const MutedText('Complete checkout before the hold expires:'),
              const SizedBox(height: Tokens.spaceSm),
              CountdownText(remainingMs),
              const SizedBox(height: Tokens.spaceLg),
              PrimaryButton('Checkout', onPressed: (reservation.isHeld && remainingMs > 0) ? onCheckout : null),
            ],
          ),
        ),
      );
}

/// 6 + 7. Order status, polled live.
class OrderStatusScreen extends StatelessWidget {
  final UiState<Order> state;
  final VoidCallback onDone;
  final VoidCallback onRetry;
  const OrderStatusScreen({required this.state, required this.onDone, required this.onRetry, super.key});
  @override
  Widget build(BuildContext context) => _scaffold(
        'Order',
        UiStateView<Order>(
          state,
          onRetry: onRetry,
          onSuccess: (order) => OrderStatusPanel(order, onDone: onDone),
        ),
      );
}
